apollo-ios/Design/rfc-caching-rewrite.md

```diff
@@ -0,0 +1,171 @@
+
+ ### Cache Overflow Handling
+
+ When an overflow of a cache is detected because it has exceeded its size limit, before any eviction takes place a call will be made to the `delegate` of the `NormalizedCacheConfiguration` to provide an opportunity to do or complete any work you feel is necessary before cache eviction takes place. You can also choose to skip the eviction, until the next check when an overflow is detected and triggered again. The `NormalizedCacheConfigurationDelegate` will look something like:
+
+ ```swift
+ public protocol NormalizedCacheConfigurationDelegate {
+  func willRunCacheEviction() -> Bool
+ }
+ ```
+
+ When this delegate function returns `true` cache eviction will take place following any configuration provided until the cache has a set amount of KB available based on what you provide with the `autoEvictionSize` in your `NormalizedCacheConfiguration`.
+
+ ### Cascading Deletions
+
+ By default when an object is delete from a cache if there are any child objects within it they will not be deleted. Cascading deletions is the idea child objects would be deleted along with the object referencing them. In order to support this there will be a parameter available on the `@cacheControl(...)` directive referenced above in the TTL sections which allows you to mark child objects for deletions in your schema. As an example given the following schema types:
+
+ ```graphql
+ type Song {
+  id: ID!
+  name: String!
+  description: String!
+  artist: Artist!
+ }
+
+ type Artist {
+  id: ID!
+  name: String!
+  genre: String!
+ }
+ ```
+
+ When deleting a `Song` from the cache, the referenced `Artist` object would not be automatically deleted unless you mark it for deletion using the directive described below.
+
+ ### Directive
+
+ As part of the `@cacheControl(...)` directive you will be able to mark child objects for deletion by marking them with `@cacheControl(cascadeDeletion: true)`. The below example shows how this would look in your schema, along with an example of marking an object for deletion while also setting its TTL:
+
+ ```graphql
+ type Song @cacheControl(maxAge: 3600) {
+  id: ID!
+  name: String!
+  description: String!
+  artist: Artist! @cacheControl(maxAge: 900, cascadeDeletion: true)
+ }
+
+ type Artist {
+  id: ID!
+  name: String!
+  genre: String!
+ }
+ ```
+
+ ### Other options
+
+ We have considered an implementation of cascading deletions where a heuristic is used to determine relationships between objects, which would detect parent/child object relationships and having the default behavior be to always delete child objects with parent objects. Currently we don't plan to move forward with this implementation.
```

Comment:
Thanks for adding the mention of this discussion. After consideration, I agree that this isn't probably something worth implementing, but I wanted a note here so that others can give us feedback on this.